iree/compiler/Conversion/LinalgToSPIRV/ConcretizeTileAmongWorkgroupsPass.cpp

```diff
@@ -0,0 +1,566 @@
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "Queried workload size: ";
+      llvm::interleaveComma(workloadSize, llvm::dbgs());
+      llvm::dbgs() << "\n";
+    });
+
+    // 3. Query the scheme for tiling among workgroups.
+
+    SmallVector<int64_t, 4> tileSize;
+    SmallVector<int64_t, 4> workgroupSize;
+
+    // Try to use configuration from the command-line first for testing.
+    tileSize.assign(options.tileSizes.begin(), options.tileSizes.end());
+    tileSize.resize(numTilableDims, 0);
+    workgroupSize.assign(options.workgroupSize.begin(),
+                         options.workgroupSize.end());
+    if (tileSize.empty() || workgroupSize.empty()) {
+      auto sizes = getTileSizeAndWorkgroupSize(rootOp, inputTypes, outputTypes);
+      if (sizes) {
+        // The tile sizes are specified against the original dimension order of
+        // the workload shape. But Flow/HAL processor id/size/count ops' are
+        // created using the reverse order.
+        tileSize = llvm::to_vector<4>(
+            llvm::reverse(sizes->first.take_front(numTilableDims)));
+        workgroupSize = llvm::to_vector<4>(sizes->second);
+      } else {
+        return funcOp.emitError("failed to query tile size and workgroup size");
+      }
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "Queried tile size: ";
+      llvm::interleaveComma(tileSize, llvm::dbgs());
+      llvm::dbgs() << "\n";
+    });
+
+    // 4. Replace hal.interface.workgroup symbolic ops with constant values.
+
+    {
+      OwningRewritePatternList patterns;
+      patterns.insert<ConcretizeWorkgroupSizeOp, ConcretizeWorkgroupCountOp>(
+          &context, workloadSize, tileSize);
+
+      (void)applyPatternsAndFoldGreedily(funcOp, std::move(patterns));
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs()
+          << "--- After concretizing hal.interface.workgroup ops ---\n";
+      funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+      llvm::dbgs() << "\n\n";
+    });
+
+    // 5. Set the entry point region for computing the number of workgroups
+    // to dispatch. The region has symbolic arguments representing the workload.
+    // So two modes here (see comments at the begining of this file).
+
+    {
+      SmallVector<int64_t, 3> numWorkgroups;
+      for (auto pair : llvm::zip(workloadSize, tileSize)) {
+        auto workload = std::get<0>(pair);
+        auto tile = std::get<1>(pair);
+        if (workload == ShapedType::kDynamicSize || tile == 0) {
+          numWorkgroups.push_back(ShapedType::kDynamicSize);
+        } else {
+          numWorkgroups.push_back(ceilDiv(workload, tile));
+        }
+      }
+
+      numWorkgroups.resize(kWorkgroupDimCount, 1);
+
+      // If all dimensions are known constant, then we can set the number of
+      // workgroups directly. Otherwise, we need to generate the IR for
+      // computing it using symbolic values.
+      if (llvm::none_of(numWorkgroups, [](int64_t dim) {
```

Comment:
I am not sure why you need this special logic for static sizes. This computation you are performing above is exactly what is done in `materializeStaticLaunchInformation`. Why can this be folded into that. Trying to avoid multiple places where the same information is set.

Reply:
IIUC, `materializeStaticLaunchInformation` computes the number of workgroups using region block arguments, like

```mlir
      hal.executable.entry_point @matmul_dynamic_shape {
      ^bb0(%arg0: index, %arg1: index, %arg2: index): 
        %c1 = constant 1 : index
        %0 = affine.apply #map0()[%arg0]
        %1 = affine.apply #map1()[%arg1]
        hal.return %0, %1, %c1 : index, index, index
      }
```

What I want here is the final constant workgroup count values, given I have both the constant problem size and constant tile size, so something like this:

```mlir
      hal.executable.entry_point @conv2d_static_shape {
      ^bb0(%arg0: index, %arg1: index, %arg2: index):
        %c2 = constant 2 : index
        %c28 = constant 28 : index
        %c28_0 = constant 28 : index
        hal.return %c2, %c28, %c28_0 : index, index, index
      }

```

So they are different branches. (And I have the comment above trying to explain this.

Reply:
I would bias towards having one place where the workgroup count is set instead of having multiple places that are doing effectively the same thing. IIUC, the workload values are from the shapes as well. So they should match up. It might be worth having access to the workload at the HAL level too I think. Something to consider. That would clean up also the need to track the output shapes, etc. 
Drop this and see if you get any correctness issues. If you do, then there is an inconsistency somewhere.

Reply:
Yeah given that we go from the same original model, host and device side should match up regarding the workload they see. Here I want the concrete constant workgroup number in device compilation, because they are used to clean up `affine.min`s in FoldGPUProcessorIDUses [patterns](https://github.com/google/iree/blob/main/iree/compiler/Conversion/LinalgToSPIRV/FoldGPUProcessorIDUses.cpp#L115). It cannot be deferred to the time when we synthesize the command buffer dispatch time in host side compilation.